service/worker/scanner/executions/workflows.go

```diff
@@ -0,0 +1,304 @@
+					report := &common.ShardScanReport{}
+					if err := workflow.ExecuteActivity(activityCtx, ScannerScanShardActivityName, ScanShardActivityParams{
+						ShardID:                 shard,
+						ExecutionsPageSize:      resolvedConfig.ExecutionsPageSize,
+						BlobstoreFlushThreshold: resolvedConfig.BlobstoreFlushThreshold,
+						InvariantCollections:    resolvedConfig.InvariantCollections,
+					}).Get(ctx, &report); err != nil {
+						errStr := err.Error()
+						shardReportChan.Send(ctx, ScanReportError{
+							Report:   nil,
+							ErrorStr: &errStr,
+						})
+						return
+					}
+					shardReportChan.Send(ctx, ScanReportError{
+						Report:   report,
+						ErrorStr: nil,
+					})
+				}
+			}
+		})
+	}
+
+	for i := 0; i < len(shards); i++ {
+		var reportErr ScanReportError
+		shardReportChan.Receive(ctx, &reportErr)
+		if reportErr.ErrorStr != nil {
+			return errors.New(*reportErr.ErrorStr)
+		}
+		aggregator.addReport(*reportErr.Report)
+	}
+
+	activityCtx = getShortActivityContext(ctx)
+	if err := workflow.ExecuteActivity(activityCtx, ScannerEmitMetricsActivityName, ScannerEmitMetricsActivityParams{
+		ShardStatusResult:     aggregator.status,
+		AggregateReportResult: aggregator.aggregation,
+	}).Get(ctx, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+// FixerWorkflow is the workflow that fixes all concrete executions from a scan output.
+func FixerWorkflow(
+	ctx workflow.Context,
+	params FixerWorkflowParams,
+) error {
+	var aggregator *shardFixResultAggregator
+	if err := workflow.SetQueryHandler(ctx, ShardReportQuery, func(shardID int) (*common.ShardFixReport, error) {
+		if aggregator == nil {
+			return nil, errQueryNotReady
+		}
+		return aggregator.getReport(shardID)
+	}); err != nil {
+		return err
+	}
+	if err := workflow.SetQueryHandler(ctx, ShardStatusQuery, func() (ShardStatusResult, error) {
+		if aggregator == nil {
+			return nil, errQueryNotReady
+		}
+		return aggregator.status, nil
+	}); err != nil {
+		return err
+	}
+	if err := workflow.SetQueryHandler(ctx, AggregateReportQuery, func() (AggregateFixReportResult, error) {
+		if aggregator == nil {
+			return AggregateFixReportResult{}, errQueryNotReady
+		}
+		return aggregator.aggregation, nil
+	}); err != nil {
+		return err
+	}
+	activityCtx := getShortActivityContext(ctx)
+	corruptedKeysResult := &FixerCorruptedKeysActivityResult{}
+	if err := workflow.ExecuteActivity(
+		activityCtx, FixerCorruptedKeysActivityName, params.FixerCorruptedKeysActivityParams).Get(ctx, &corruptedKeysResult); err != nil {
+		return err
+	}
+
+	resolvedConfig := resolveFixerConfig(params.FixerWorkflowConfigOverwrites)
+	shardReportChan := workflow.GetSignalChannel(ctx, fixShardReportChan)
+	for i := 0; i < resolvedConfig.Concurrency; i++ {
+		idx := i
+		workflow.Go(ctx, func(ctx workflow.Context) {
+			for j, corrupted := range corruptedKeysResult.CorruptedKeys {
+				if j%resolvedConfig.Concurrency == idx {
+					activityCtx = getLongActivityContext(ctx)
```

Comment:
I think if the scanner workflow is still running, that error should be treated as non-retryable?

Reply:
Which line is this referring to? If this is referring to line 272 and beyond then at this point we already are sure the scan workflow is closed that is checked in FixerCorruptedKeysActivity (which is short running so if the workflow is still running it will return quickly).

Reply:
I should have commented in `activities.go` L219. There's a check for the workflow close status and return error if close status == nil. I think we should not retry in that case?

Reply:
Synced offline. Will be addressed in next pr.